Incorrect handling of some PNG pixel formats in WebGPU

mwyrzykowski opened the pull request:
#### 190091a8162fff720f5379d28ee2aa298c646127
<pre>
Incorrect handling of some PNG pixel formats in WebGPU
<a href="https://bugs.webkit.org/show_bug.cgi?id=297816">https://bugs.webkit.org/show_bug.cgi?id=297816</a>
<a href="https://rdar.apple.com/158797747">rdar://158797747</a>

Reviewed by Tadeu Zagallo.

We incorrectly handled grayscale, alpha-only, and gray+alpha imports
along with 16 bits per channel images.

* Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
(WebCore::clipTo8bpp):
(WebCore::imageBytesForSource):

Canonical link: <a href="https://commits.webkit.org/299404@main">https://commits.webkit.org/299404@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/adc9bceb0b4c291d67941c28c113c79c5ff5e6af

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/118908 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/38589 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29240 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125102 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/70971 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/bf66f4fa-3374-42c8-9287-8bd957d3dabf) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/120786 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39285 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47171 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90240 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/59757 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/b6486f57-ec62-470b-8fc6-ced43030511c) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/121861 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31318 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/106600 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70745 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/6293f080-297a-4266-8913-bbc41ffb88e1) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30381 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/24714 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/68760 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/100764 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24902 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128148 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/45815 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/34607 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98897 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46181 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102820 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/98677 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25081 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44128 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/22131 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/42380 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/45685 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/51363 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45151 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/48495 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/46835 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

tadeuzagallo commented on Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
```diff
@@ -411,6 +411,25 @@ static void getImageBytesFromVideoFrame(WebGPU::Queue& backing, const RefPtr<Vid
 #endif
 
+#if PLATFORM(COCOA) && ENABLE(VIDEO) && ENABLE(WEB_CODECS)
+static void clipTo8bpc(std::span<const uint8_t> data, size_t bitsPerComponent, Vector<uint8_t>& byteSpanBacking)
```
Just a nit, but maybe it's worth avoiding the `bpc` abbreviation?

mwyrzykowski replied:
good idea, expanded to `bitsPerChannel`

mwyrzykowski committed:
Incorrect handling of some PNG pixel formats in WebGPU
https://bugs.webkit.org/show_bug.cgi?id=297816
rdar://158797747

Reviewed by Tadeu Zagallo.

We incorrectly handled grayscale, alpha-only, and gray+alpha imports
along with 16 bits per channel images.

* Source/WebCore/Modules/WebGPU/GPUQueue.cpp:
(WebCore::clipTo8bpp):
(WebCore::imageBytesForSource):

Canonical link: https://commits.webkit.org/299404@main
